kong/api/routes/health.lua

```diff
@@ -14,6 +15,20 @@ local kong = kong
 
 
 return {
+  ["/config/ready"] = {
+    GET = function(self, dao, helpers)
+      if kong.db.strategy ~= "off" then
+        return kong.response.exit(200)
+      end
+    -- unintuitively, "true" is unitialized. we do always initialize the shdict key
+    -- after a config loads, this returns the hash string
+    if not declarative.has_config() then
```

Comment:
It means `declarative.load_into_cache()` has completed. If you're running strict consistency, the router and such will be updated since they'll rebuild at request time. If eventual, it's rebuilt whenever that fires (maybe? not sure if that system behaves differently in DB-less).

For the controller's purposes that's fine, since we mainly just care about knowing when we need to push it again.

For general purpose config availability, I don't know if there's a good approach for knowing when the config is available, since it's built per-worker--I can set an additional SHM key, but it won't be guaranteed that all workers have rebuilt. Is there anything we can use to inform us when all workers have completed processing of the declarative flip_config event?

Reply:
c3fef049d868fd0d1dc98847ea8db6b090e6561a sorta does this, but it only means that at least one worker has rebuilt using the new config.

Reply:
I don't think there is any plumbing for what I'm requesting, is there @bungle?

Reply:
I confirm that what @rainest says is correct: his flag indicates that the declarative config might have been loaded into the node's memory, but it can take a while until all the nodes get their routers (and plugins iterators) updated. With strict consistency set, requests "will wait" if a Router is not up to date so for all effects and purposes the new exposed flag it will indicate that "Kong will use the new config, even if it means waiting". 

With eventual consistency, the gateway itself "does not care" about exact availability. If something else cares, then I think the answer to `the router and run loops have been updated to use the config` can't be a single bit. Consider that some customers update their declarative configurations very rapidly. Workers are in config A. Config B arrives, gets processed by 1 worker, which sets the flag. Workers gradually load config B. While this is happening, Config C arrives. And so on. In that scenario with frequent updates, a single bit will always be `false` - a number of workers will always be out of date.

Options:
* Change the question to `is the router and and run loops updated for this specific config (or a more recent one)?`. We could answer a bit for that one.
* Change the question to `how many workers are updated to this specific config`. The answer to that one would be a number, or a couple numbers (2 out of 16).
* Change the question to `which config does every worker have`. The answer would be a list of n configuration identifiers. Probably their hashes.

I don't think we have any of the plumbing for that at the moment.

Reply:
Do we have something that provides the worker counts/a list of their PIDs? It's more complicated, but it should be viable to restructure the value of `DECLARATIVE_CONFIG_READY_KEY` from a bit to N bits, and then in the endpoint loop over N to confirm that all are set.

We're not really concerned with latest config, just that configuration is not null, so the update stampede case isn't a problem. You just always set your worker entry to `true` on update (there's probably some way to make this only happen once using local worker memory to avoid the unnecessary SHM locks after the first update--is there a Lua equivalent of https://pkg.go.dev/sync#Once?) and don't worry what specific hash you've loaded.

Reply:
ed5d5d736 updates the previous system to add a worker PID onto the end, so we set and then check a key per worker, along with a better test to confirm that it works as expected. We can't test actual startup conditions easily (at all?) AFAIK, but most of the moving parts are in the bit where workers mark their status.

Reply:
I'm not the SME here so I'll defer the review to someone within Gateway team.
Health-endpoints are invoked by schedulers, orchestrators, load-balancers, external health-checkers, etc. Meaning, they should be very performant. The shm:gets acquire a lock across all workers, which can drastically affect performance.